Draft EIP: Rename `gas` to `mana` (#6789)
ethereum/EIPs: EIPS/eip-6789.md

---
eip: 6789
title: Rename gas to mana
description: This EIP suggests renaming gas to mana, as proposed by Vitalik Buterin in 2015

> author: Pascal Caversaccio (@pcaversaccio)

Reviewer: IMO Vitalik should have been listed as co-author as the idea originated from him.

Reviewer 2: It's a bit of a gray area. I would like him to be listed as a co-author, but I also feel that he should explicitly consent to being a co-author.

CC @vbuterin in case he responds.

---
eip: 6789
title: Rename `gas` to `mana`
description: This EIP picks up on Vitalik Buterin's (@vbuterin) original proposal in 2015 to rename `gas` to `mana`
author: Pascal Caversaccio (@pcaversaccio)
discussions-to: https://ethereum-magicians.org/t/eip-6789-rename-gas-to-mana/13570
status: Draft

> type: Standards Track
> category: Interface

Reviewer 2: ```suggestion
type: Standards Track
category: Core
```
I was trying to determine whether to have it be Meta, Informational, or Core. I think Core most accurately describes this EIP, but I encourage other editors to chime in.

Author: Here we go again :). That was also my first choice, but Matt had another opinion here https://github.com/ethereum/EIPs/pull/6789#discussion_r1149559967. It seems my EIPs are always a challenge to assign a category. So, how should we proceed here?

Reviewer 2: I would say that Interface is *definitely* not the right category for this.

Author: Fully agreed - I read the [EIP-1](https://eips.ethereum.org/EIPS/eip-1) again and it states:

> **Core**: [..] as well as changes that are not necessarily consensus critical but may be relevant to [“core dev” discussions](https://github.com/ethereum/pm) (for example, [EIP-90], [...]

=> This is definitely critical at some point to the "core dev" discussions.

Reviewer: @Pandapip1 the precedence is clearly to use `Interface` for renaming mnemonics and the like. See [EIP-6](https://eips.ethereum.org/EIPS/eip-6).

Reviewer 2: Okay, if there's a precedent, I'm okay with interface.

Author: Resolving this discussion for the moment.

created: 2023-03-27
---

## Abstract

> This EIP picks up on Vitalik Buterin's (@vbuterin) original proposal in 2015 to rename `gas` to `mana`.

Reviewer 2: ```suggestion
This EIP suggests renaming `gas` to `mana`, as proposed by Vitalik Buterin in 2015.
```

> ## Motivation

Reviewer 2: Additionally, you seem to have confused the Motivation and Rationale sections. The text you have here is great Rationale, and the text you have in the Rationale is great Motivation.

> Vitalik's original motivation is threefold and is still valid today:
> 
> - `mana` more accurately reflects the increased environmental friendliness of Proof-of-Stake (PoS),
> - `mana` is generally understood to be ephemeral and non-transferable, and thus represents accurately the characteristic of how `gas` is actually used within Ethereum, and
> - `mana` is generally considered renewable, whilst (natural) gas is considered non-renewable.

Reviewer 2: ```suggestion
- `mana` reflects the increased environmental friendliness of Proof-of-Stake;
- `mana` is generally understood to be ephemeral and non-transferable, which better represents the concept of `gas`; and
- `mana` is generally portrayed as renewable, while (natural) `gas` is non-renewable.
```

Author: FWIW, I abbreviated PoS since I re-used it below in the rationale, but since am gonna switch both sections anyways I will use your suggestion.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> The core term `gas` MUST be renamed to `mana`. This MUST also include renaming the opcode `Gas` to `Mana`.

Reviewer 2: ```suggestion
The core term `gas` MUST be renamed to `mana`.

The following opcodes MUST be renamed:

- `GASPRICE` to `MANAPRICE`;
- `GASLIMIT` to `MANALIMIT`;
- `GAS` to `MANA`

Additionally, the parameters/outputs of the following opcodes MUST be renamed:

- `CALL`'s `gas` parameter to `mana`
- `CALLCODE`'s `gas` parameter to `mana`
- `DELEGATECALL`'s `gas` parameter to `mana`
- `STATICCALL`'s `gas` parameter to `mana`
- `GASLIMIT`'s `gasLimit` output to `manaLimit`
- `GAS`'s `gas` output to `mana`

Finally, the following RPC endpoints MUST be renamed:

- `eth_estimateGas` to `eth_estimateMana`
- `eth_gasPrice` to `eth_manaPrice`
- `eth_maxPriorityFeePerGas` to `eth_maxPriorityFeePerMana`
```
I think this specification is more complete.

Author: Yes fully agreed. I was not sure in first place, whether to use an _implied_ specs or an explicit one. This one is much more clear! Thx.

## Rationale

The underlying rationale for reviving Vitalik's original proposal from 2015 is that we have finally arrived at the age of PoS, and given the roadmap ahead (i.e. "The Surge", "The Scourge", "The Verge", "The Purge", "The Splurge"), I consider this moment as the last opportunity to make such a far-reaching semantic change.

## Backwards Compatibility

This proposal is not backward compatible as it renames the core term `gas`.

## Test Cases

### Example 1

If a transaction requires more `mana` than allowed by the `manaLimit`, it is reverted as an _out-of-mana_ transaction.

### Example 2

A Solidity contract to estimate the used `mana` via the new `manaleft()` syntax (replacing `gasleft()`) for dedicated function calls.

```solidity
// SPDX-License-Identifier: AGPL-3.0
pragma solidity 0.8.19;

contract ManaMetering {
    function oldWay() external view returns (string memory, uint256 manaUsed) {
        string memory hiMom = "Hi Mom, ";
        string memory missYou = "miss you.";

>         uint256 startMana = manaleft();

Reviewer 3: Since we are in the oldWay function I think you wanted to say ``startGas`` instead of ``startMana``.
Both Example are similar.

        string memory concat = string(abi.encodePacked(hiMom, missYou));
        manaUsed = startMana - manaleft();
        return (concat, manaUsed);
    }

    function newWay() external view returns (string memory, uint256 manaUsed) {
        string memory hiMom = "Hi Mom, ";
        string memory missYou = "miss you.";
        uint256 startMana = manaleft();
        string memory concat = string.concat(hiMom, missYou);
        manaUsed = startMana - manaleft();
        return (concat, manaUsed);
    }
}
```

In Vyper, the same behaviour can be replicated with the new transaction property `msg.mana`, which replaces `msg.gas`.

### Example 3

An example of how to set the `manaLimit` in MetaMask:

![MetaMask manaLimit](../assets/eip-6789/MetaMask_ManaLimit.png)

> ## Security Considerations

Reviewer 2: What about renaming the RPC endpoints? Could that potentially have consequences?

Author: In https://github.com/ethereum/EIPs/pull/6789/commits/00c1865164587616ad4f889f29a7c534f07849a4 I added further notes on how the RPC interpretations change. The value itself is not changed. Thus, if a library such as `ethers` calls `eth_maxPriorityFeePerGas` after the change, either it will still work since an alias was introduced or it will revert with a message that the method is not available. The change breaks a lot of communication channels and endpoints, but the value itself and how it's used are untouched. That's why I don't see any direct security risks.
